server/build_event_protocol/event_parser/event_parser.go

```diff
@@ -206,8 +214,10 @@ func (sep *StreamingEventParser) FillInvocation(invocation *inpb.Invocation) {
 		buildDuration = time.Duration((sep.endTimeMillis - sep.startTimeMillis) * int64(time.Millisecond))
 	}
 	invocation.DurationUsec = buildDuration.Microseconds()
-	// TODO(siggisim): Do this rendering once on write, rather than on every read.
-	invocation.ConsoleBuffer = string(sep.screenWriter.RenderAsANSI())
+	if sep.screenWriter != nil {
+		// TODO(siggisim): Do this rendering once on write, rather than on every read.
+		invocation.ConsoleBuffer = string(sep.screenWriter.RenderAsANSI())
+	}
```

Comment:
In order for screenWriter to have nonempty contents at this point, it looks like this `FillInvocation` call needs to happen twice, once initially, and again after parsing all events, which is what we currently do when calling `fillInvocationFromEvents`.

IMO this is a fairly error-prone API, and ideally we'd refactor at some point in the future so that `FillInvocation` only needs to be called once after parsing all events. So, if possible, it'd be nice to not add more dependencies on calling `FillInvocation` twice. Would it make sense to instead pass an optional screenWriter into `NewEventParser`?

Reply:
What I mean is, in `fillInvocationFromEvents`, we do this (pseudo-code):

```go

parser := NewParser()
parser.FillInvocation(inv) // screenWriter is initialized here but nothing is written to it
for event := events {
  parser.Parse(event) // stuff gets written to screenWriter
}
parser.FillInvocation(inv) // contents of the existing screenWriter get written to consoleBuffer

```

I think it would be more clear if screenWriter were initialized in NewParser(), if possible. Currently we depend on `FillInvocation` being called before all events are parsed, but I think eventually it would be nice to remove that dependency.

Reply:
Oh! Now I understand. Yeah, you're right, that is a problem. Fixed!